### DISCUSSION TITLE
Hasil Deploy dengan run App shinny berbeda

### DISCUSSION THREAD

**Discussion by DeanAlexander27** [teamalgoritma/community#577] [Q&A] [answered]
Dear Ka Instruktur
Saya mengalami kendala ketika hasil deploy dengan run app shinny berbeda sebagai berikut
![image](https://user-images.githubusercontent.com/123348110/221433449-731a8240-d4df-447b-8336-8e39d6b3084f.png)

![image](https://user-images.githubusercontent.com/123348110/221433471-d7446323-8fb2-41b0-9a9e-9d1aea962290.png)

memang di basic data cleansing saya membenarkan nama negara dengan ASCII dan UTF8 dengan iconv tetapi ketika run app nya berjalan normal hanya terkendala di deploy..

terima kasih

**Comment by idoali** [comment]
Terima kasih Pak Dean atas pertanyaannya

Maaf pak untuk kesalahannya apa boleh diperjelas lagi? Untuk gambar pertama apakah salahnya karena Republik Congo dituliskan dengan nama yang tidak sesuai? Untuk plot kedua apakah kesalahannya karena ada bar yang tidak semestinya ada tetapi justru ada?

Jika ada hasil yang berbeda ketika dideploy, ini adalah kejadian yang jarang pak. Apalagi jika kesalahannya terdapat pada plot Plotly (yang sering terjadi adalah di bagian tags HTML, tetapi kami berasumsi bapak tidak menggunakan itu). Solusi dari kami sementara ini adalah mendeploy ulang program bapak.

**Reply by DeanAlexander27** [reply]
halo Ka Ido
iya ,, jadi saat saya run app seperti biasa Republik Congo itu muncul, tetapi ketika di deploy dan keluar di web shinnyapps, tidak muncul ..
sebelumnya data set ini agak cukup berantakan karena nama negaranya tidak beraturan saya menggunakan iconv ASCII dan UTF8

iya per pagi ini juga sudah deploy 3x hasilnya tetap sama. saya cukup kebingungan untuk case seperti ini
![image](https://user-images.githubusercontent.com/123348110/221476676-32a65f75-2d73-4e92-a2c4-5ee7900b389c.png)

seperti mencari republic congo dan republic lainnya tidak ada.. sudah diset ke factor dari server.R nya tapi hasilnya tetap sama.. terima kasih

**Comment by idoali** [comment]
Terima kasih pak atas konfirmasinya.

Ada dua cara yang bisa dilakukan disini pak : 
1. Coba pakai fungsi `wrap_format()` dari library `scales`. Dengan fungsi ini kita bisa menampilkan tulisan yang panjang menjadi beberapa baris. Fungsi ini sendiri kita masukkan saat membuat ggplot. Contoh penggunaannya adalah seperti ini pak : 

```{R}
ggplot(data = data, aes(x = x, y = y)) +
  geom_col(show.legend = F) +
  scale_y_discrete(labels = wrap_format(25))
```

Dimana angka 25 diatas adalah jumlah karakter yang akan kita tampilkan di satu baris

2. Jika masih tidak terlihat, "Republic Congo" bisa diganti saja menjadi "Congo" pak, agar nama menjadi lebih singkat. Untuk melakukan ini bapak bisa memanfaatkan fungsi `ifelse()` atau fungsi-fungsi yang lain.

**Reply by DeanAlexander27** [reply]
ah baik ka Ido saya coba jalankan.. mohon untuk capstone ini bisa saya lampirkan yang terupdate langsung di shinny nya ya ka. terima kasih

**Reply by DeanAlexander27** [reply]
dear ka Ido
saya sudah menemukan beberapa bottle neck ,seperti nama mixico , dll yang penulisan negara tidak sesuai dengan penulisan pada umumnya, saya sudah pakai iconv untuk ke utcf8 dan ascii ...

secara run app berjalan normal, tetapi hasil deploy ketika melakukan ifelse rep congo menjadi NA dan dibaca menjadi 1 negara bernama NA dari rep congo,mixico, dll sehingga terbaca NA paling mendominasi dari seluruh kategori..
![image](https://user-images.githubusercontent.com/123348110/221599212-67c603ac-87f3-42a0-be79-1b371499a3ab.png)

apakah ada solusi untuk masalah ini ya ? 

terima kasih

berikut codingannya 

dataset <- read_csv("DataCoSupplyChainDataset.csv")
dataset <- dataset %>% 
  mutate_at(.vars=c("Type", "Delivery Status", "Category Id","Category Name","Customer Segment","Customer State","Customer Street","Customer Zipcode","Customer City","Customer Country","Department Id","Department Name","Market","Order City","Order Country","Customer Id","Order Region","Order State","Order Status","Product Name","Shipping Mode"),as.factor) %>% mutate_all(funs(replace(., is.na(.), 0))) %>% 
  select(-c(`Customer Email`,`Customer Password`)) %>% 
  mutate(`order date (DateOrders)` = mdy_hm(`order date (DateOrders)`)) %>% 
  mutate(`shipping date (DateOrders)` = mdy_hm(`shipping date (DateOrders)`)) %>% 
  mutate(date1 = as.Date(`order date (DateOrders)`,format = "%d/%m/%Y")) %>%
  mutate(date2 = as.Date(`shipping date (DateOrders)`,format = "%d/%m/%Y")) %>% 
  mutate(month_year1 = as.factor(format(date1, "%Y/%m"))) %>%
  mutate(month_year2 = as.factor(format(date2, "%Y/%m"))) %>% 
  mutate(year = year(date1)) %>% 
  mutate(quarter = quarter(date1)) %>% 
  mutate(year_quarter = as.factor(paste(year, "Q", quarter))) %>% 
  mutate(`Order Country` = iconv(`Order Country`,"ASCII","UTF-8")) %>%
  mutate(`Order Country` = ifelse(`Order Country` %in% c("Repzblica Democratica del Congo", "Repzblica Dominicana", "Repzblica Checa", "Repzblica del Congo", "Repzblica Centroafricana", "Repzblica de Gambia"), 
                                  substring(`Order Country`, 11),
                                  `Order Country`)) %>%
  mutate(`Order Country` = as.factor(`Order Country`))

**Comment by idoali** [comment, marked as answer]
Baik pak

Contoh penggunaan ifelse seperti ini pak : 

```{r}
# ifelse(kondisi, nilai jika kondisi benar, nilai jika kondisi salah)
ifelse(country == "Repzblica Democratica del Congo", "Congo", country)
```

Fungsi `ifelse()` sendiri hanya bisa kita gunakan untuk mengubah berdasarkan satu kondisi saja pak. Pada kasus bapak sendiri, kondisinya terlalu banyak (mengubah nama congo, dominic, checa dll). 

Jika kasusnya mengandung kondisi yang banyak, kita bisa mengganti ifelse dengan `switch()`. Ini sudah pernah dipraktikkan di kelas Data Visualization pak. Contoh penggunaannya adalah seperti ini : 

```{r}
y <- switch(x,
              "1" = "Film and Animation",
             "2" = "Autos and Vehicles", 
             "10" = "Music", 
             "15" = "Pets and Animals", 
             "17" = "Sports",
             "19" = "Travel and Events", 
             "20" = "Gaming", 
             "22" = "People and Blogs", 
             "23" = "Comedy",
             "24" = "Entertainment", 
             "25" = "News and Politics",
             "26" = "Howto and Style", 
             "27" = "Education",
             "28" = "Science and Technology", 
             "29" = "Nonprofit and Activism",
             "43" = "Shows")
```

**Reply by DeanAlexander27** [reply]
Baik Ka ido.. terima kasih segera saya uji coba kan

**Reply by DeanAlexander27** [reply]
terima kasih ka @idoali dan ka @kevbow 
masalah sudah ter solved .
hanya perlu di export lagi data yang sudah di cleansing, dan digunakan kembali
